# Анализ комбинаторных задач с буквами: система счисления и ограничения

## Постановка задачи

Все пятибуквенные слова, составленные из русских букв **Я, Н, В, А, Р, Ь**, записаны в алфавитном порядке и пронумерованы, начиная с 1.

**Начало списка:**
1. ААААА
2. ААААВ
3. ААААН
4. ААААР
5. ААААЬ
6. ААААЯ
7. АААВА
...

Под каким номером в списке идёт последнее слово, которое не начинается с буквы Я, содержит не более одной буквы Ь и не содержит букв Я, стоящих рядом?

---

## Методы решения

Рассмотрим 3 подхода к решению:
1. **Ручное решение** (через систему счисления)
2. **Решение с вложенными циклами**
3. **Решение с помощью библиотеки itertools**

---

## 1. Ручное решение через систему счисления

!> **ВАЖНО:** Данное решение предназначено именно для вопроса про номер конкретной комбинации. Так получается, что у букв этот вопрос встречается чаще, чем у цифр.

!> **ВАЖНО:** Буквы должны быть обязательно расположены в алфавитном порядке, в условии они даны произвольным образом.

### Определение алфавитного порядка

Можно начать вспоминать алфавит, но авторы задачи уже сделали всё за нас - нужно быть внимательнее!

**Обратите внимание на пример списка, который нам даётся:**

Заметили? Автор специально дал в примере самые первые комбинации, благодаря чему по последним буквам можно определить алфавитный порядок.

![Последовательность букв](images/1.png)

Из анализа начала списка видно, что буквы идут в порядке: **А → В → Н → Р → Ь → Я**

!> **ОСОБОЕ ВНИМАНИЕ:** В условии указано, что буквы **Я не должны стоять рядом друг с другом**, но говорится о том, что буквы Я не должны стоять рядом друг с другом. Вот такая путаница в формулировках!

### Кодирование букв цифрами

Для решения мы закодируем наши буквы под систему счисления. Для того чтобы было более удобно составлять комбинацию, а так же посчитать сразу номер искомой комбинации.

**Кодируем следующим образом:**

![Кодирование букв](images/2.png)

**Схема кодирования:**
- А → 0
- В → 1  
- Н → 2
- Р → 3
- Ь → 4
- Я → 5

### Построение максимальной допустимой комбинации

Поскольку от нас требуют найти максимальную комбинацию, то нам нужно составить из нашей системы счисления максимально возможную комбинацию по числовому значению, при этом соблюдая оставшиеся условия.

Итак, у нас сказано в начале задачи, что у нас 5-буквенные слова, значит у нас 5 пустых ячеек для составления комбинаций:

![Пустые позиции](images/3.png)

#### Шаг 1: Первая ячейка
Смотрим в условие задачи и видим, что у нас в начале не может стоять буква Я(5), которая имеет самое высокое числовое значение среди всех букв. Значит мы должны поставить следующую максимальную букву - это буква Ь(4), она имеет числовое значение 4.

![Позиция 1](images/4.png)

#### Шаг 2: Вторая ячейка
Ограничений по второй ячейке у нас нет, значит ставим туда букву с самым большим числовым значением. Это буква Я(5).

![Позиция 2](images/5.png)

#### Шаг 3: Третья ячейка
По третьей ячейке тоже ограничений нет, значит ставим снова самую большую букву по числовому значению Я(5). Но не тут-то было! По условию буквы Я не могут стоять рядом. Значит ставим следующую букву по максимальному значению Ь(4). И снова мимо - по условию у нас сказано, что букв Ь должно быть не более одной, а мы уже использовали её в первой ячейке. Значит следующая буква по максимальному числовому значению - это Р(3).

![Позиция 3](images/6.png)

#### Шаг 4: Четвертая ячейка
Ставим максимальную букву по числовому значению Я(5). Рядом нет букв Я(5), так что мы можем это сделать.

![Позиция 4](images/7.png)

#### Шаг 5: Пятая ячейка
Тут такая же история, как и в третьей ячейке: рядом уже стоит Я(5), а Ь(4) мы уже использовали. Значит ставим Р(3).

![Позиция 5](images/8.png)

### Перевод из шестеричной системы в десятичную

И вот ответ в наших руках! Но давайте вспомним, что мы кодировали наши буквы в цифры, и по сути мы закодировали наши буквы в шестеричную систему счисления. Значит ответ мы получили не в привычной десятичной системе, а в шестеричной - надо это исправить.

**Переведём получившееся число 45353 из шестеричной системы в десятичную:**

1) Расставляем справа налево индексы над ячейками:

![Разряды числа](images/9.png)

2) Теперь берём слева направо числа и умножаем его на основание системы счисления (6), возведенное в степень, которая стояла над умножаемым числом, и складываем результаты:

![Вычисление](images/10.png)

**Считаем с помощью Python:**

In [ ]:
x1 = 4*6**4 + 5*6**3 + 3*6**2 + 5*6**1 + 3*6**0
print(x1)

6405


### Проверка решения и важное замечание

Получили число **6405** - и вот оно, счастье! Но не тут-то было. 

Сравнив ответы, мы обнаруживаем, что правильный ответ больше ровно на **1**. 

!> **КРИТИЧЕСКИ ВАЖНО:** Почему произошла эта ошибка? Ошиблись в расчётах? Неправильно перевели?

**Всё проще:** то, что мы перевели в десятичную систему - это номер комбинации, если бы отсчёт начинался с **0**. Но в нашей задаче отсчёт начинается с **привычной 1**!

!> **ПРАВИЛО:** При переводе позиции комбинации из любой системы счисления в номер в списке (с нумерацией с 1) необходимо всегда прибавлять **+1** к полученному результату.

**Итоговый расчёт:**

In [ ]:
x1+1

6406

In [ ]:
# А можно немного схитрить!
# Правильный расчет номера слова
number_6 = "45353"    # Наша комбинация в 6-ричной системе
number_10 = int(number_6, 6)  # Перевод в десятичную систему
final_answer = number_10 + 1   # Учитываем нумерацию с 1
print(final_answer)

6406


# Программные решения задачи

## 2. Решение с помощью вложенных циклов

In [ ]:
# Счетчик для нумерации всех слов (включая неподходящие)
count = 0

# Алфавит букв в порядке следования
letters = 'АВНРЬЯ'

# Список для хранения номеров подходящих слов
ans = []

# Первый цикл - первая буква слова
# Используем только 'АВНРЬ' (исключаем 'Я'), так как слово не должно начинаться с Я
for b1 in 'АВНРЬ':
    # Второй цикл - вторая буква (может быть любой из алфавита)
    for b2 in letters:
        # Третий цикл - третья буква
        for b3 in letters:
            # Четвертый цикл - четвертая буква
            for b4 in letters:
                # Пятый цикл - пятая буква
                for b5 in letters:
                    # Увеличиваем счетчик (номер текущего слова)
                    count += 1
                    
                    # Формируем слово из пяти букв
                    word = b1 + b2 + b3 + b4 + b5
                    
                    # Проверяем условия:
                    # 1. Не более одной буквы Ь (0 или 1)
                    # 2. Нет двух букв Я подряд
                    if word.count('Ь') <= 1 and 'ЯЯ' not in word:
                        # Если условия выполнены, добавляем номер слова в список
                        ans.append(count)

# Выводим последний элемент списка - номер последнего подходящего слова
print(ans[-1])

## Версия кода без комментариев

In [13]:
count = 0
letters = 'АВНРЬЯ'
ans = []
for b1 in 'АВНРЬ':
    for b2 in letters:
        for b3 in letters:
            for b4 in letters:
                for b5 in letters:
                    count += 1
                    word = b1 + b2 + b3 + b4 + b5
                    if word.count('Ь') <= 1 and 'ЯЯ' not in word:
                        ans.append(count)
print(ans[-1])


6406


## 3. Решение с помощью библиотеки itertools

In [21]:
# Импортируем все функции из модуля itertools
from itertools import *

# Задаем алфавит - разрешенные буквы в алфавитном порядке
alp = 'АВНРЬЯ'

# Генерируем все возможные комбинации длиной 5 из букв алфавита
# product() создает декартово произведение, repeat=5 означает 5 позиций
combs = product(alp, repeat=5)

# Счетчик для нумерации всех слов (включая неподходящие)
count = 0

# Список для хранения номеров подходящих слов
ans = []

# Перебираем все сгенерированные комбинации
for comb in combs:
    # Увеличиваем счетчик (номер текущего слова)
    count += 1
    
    # Преобразуем кортеж символов в строку
    # Например: ('А','А','А','А','А') → "ААААА"
    comb = ''.join(comb)
    
    # Проверяем первое условие: слово не начинается с буквы Я
    if comb[0] not in ['Я']:
        # Проверяем остальные условия:
        # 1. Не более одной буквы Ь (0 или 1)
        # 2. Нет двух букв Я подряд
        if comb.count('Ь') <= 1 and 'ЯЯ' not in comb:
            # Если все условия выполнены, добавляем номер слова в список
            ans.append(count)
            
# Выводим последний элемент списка - номер последнего подходящего слова
print(ans[-1])

6406


## Версия кода без комментариев

In [22]:
from itertools import *
alp = 'АВНРЬЯ'
combs = product(alp,repeat=5)
count = 0
ans = []
for comb in combs:
    count +=1
    comb = ''.join(comb)
    if comb[0] not in ['Я']:
        if comb.count('Ь') <= 1 and 'ЯЯ' not in comb:
            ans.append(count)
            
print(ans[-1])

6406


## Рефлексия

### 🔍 Что узнали:
- **Анализ условий**: определяем порядок букв по началу списка
- **Системы счисления**: кодируем буквы в цифры для удобства расчетов  
- **Нумерация**: всегда учитываем разницу между отсчетом с 0 и с 1
- **Множественные подходы**: ручной расчет, вложенные циклы, itertools

### 💡 Главные выводы:
- Всегда проверяйте начало нумерации в условии
- Сочетание ручного и программного решения дает надежный результат
- Itertools удобен для проверки, но может быть медленным для больших данных

**Итог**: освоили 3 метода решения комбинаторных задач с буквами